### PR TITLE
Add missing types and parameters to Notifications API

### LINKED ISSUE #313: Add missing types and parameters to Notifications API
We already implemented the [notifications API](https://docs.joinmastodon.org/methods/notifications/) but recent versions of Mastodon added new types and query parameters which we should add.

## “Get all notifications” query parameters

We offer `excludeTypes` but not `types` which can be used to _include_ types.

`account_id` is also missing.

* [x] Add optional `types` query parameter as `List<Notification.Type>?` with default value `null`
* [x] Add optional `account_id` query parameter as `String` with default value `null`

## Notification entity properties

We currently only support four report types

* [x] Add `status` = Someone you enabled notifications for has posted a status
* [x] Add `follow_request` = Someone requested to follow you
* [x] Add `poll` = A poll you have voted in or created has ended
* [x] Add `update` = A status you interacted with has been edited
* [x] Add `admin.sign_up` = Someone signed up (optionally sent to admins)
* [x] Add `admin.report` = A new report has been filed

> This type is missing `status`, `follow_request`, `poll`, `update`, `admin.sign_up`, and `admin.report` but we should probably only add them in a follow-up PR where we check all notification types again.

_Originally posted by @PattaFeuFeu in https://github.com/andregasser/bigbone/pull/310#discussion_r1375989169_

---

**Please only start with this once #310 has been merged!**

### PR DESCRIPTION
**Pull request by PattaFeuFeu**
# Description

This PR adds missing parameters `types` and `account_id` to the [API to get all notifications](https://docs.joinmastodon.org/methods/notifications/#get).

It also adds missing `NotificationType`s.

To better handle future updates of the Mastodon API where we aren’t fast enough to release a new version, I have changed the fallback value to `null` instead of `MENTION`.

I have changed the `Parameters` class a bit to improve testability. All unit tests still pass and outcome is as expected.

For review, it could make sense to go commit by commit.

Closes #313 

# Type of Change

- New feature (non-breaking change which adds functionality)
- Breaking change
  - `type` is now nullable
- This change requires a documentation update
  - [API Coverage](https://github.com/andregasser/bigbone/wiki/Mastodon-API-Coverage) page needs to be updated

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. 

# Mandatory Checklist

- [x] My change follows the projects coding style
- [x] I ran `gradle check` and there were no errors reported
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] KDoc added to all public methods
  - I have not added KDocs for methods I didn’t change. Should better be done as part of #318.

### CODE REVIEW

File: bigbone/src/main/kotlin/social/bigbone/api/entity/Notification.kt [hunk 1]
```diff
@@ -54,16 +54,41 @@ data class Notification(
...
-        FAVOURITE,
+        @SerialName("status")
+        STATUS,
 
-        @SerialName("follow")
-        FOLLOW
+        @SerialName("update")
+        UPDATE;
+
+        open val apiName: String = name.lowercase()
```
Comment (PattaFeuFeu) [inline bigbone/src/main/kotlin/social/bigbone/api/entity/Notification.kt:92]: I had to add `apiName` now after all… 😞 In #310 we thought that we’d only need the api names for tests and decided it’s not worth it to keep the `apiName` parameter up-to-date for that.

But we also need the `NotificationType` for the query parameters `types` and `exclude_types` of the “get all notifications” method. `admin.report` and `admin.sign_up` both contain dots which are not allowed in enum names.

For me, this was the best solution here. If you have a better idea, feel free to comment.

Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/api/entity/Notification.kt:92]: Ah, wait. Found a better way! 🥳 

https://github.com/Kotlin/kotlinx.serialization/issues/1482#issuecomment-841308774

Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/api/entity/Notification.kt]: I did not update other classes as that would be out of scope for this PR.

If you think this is a good way forward, I’m happy to create a ticket.

**Comment by PattaFeuFeu** [comment]
Could we move forward with this one here? 😊 @andregasser @bocops

File: bigbone/src/main/kotlin/social/bigbone/Parameters.kt [hunk 2]
```diff
@@ -18,7 +21,7 @@ class Parameters {
      * @return this Parameters instance
      */
     fun append(key: String, value: String): Parameters {
-        parameterList.add(Pair(key, value))
+        parameters.getOrPut(key, ::mutableListOf).add(value)
```
[the remark this answers is not in the document]
Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/Parameters.kt:24]: `parameters.getOrPut(key, lambda)` can be read as follows:

* Get the `key` in parameters if available
* If it is not available, run the code in the lambda to add to `parameters` and _then_ return it, if necessary

So what this does:

Get `key` in parameters` and add `value` into it.
If `key` doesn’t exist, first create a new empty `MutableList`, put it into `parameters` as `key`, and then add `value` to _that_ one.

**Review by andregasser — DISMISSED** [review]
Thanks, looks good to me! 👍

**Comment by PattaFeuFeu** [comment]
Again please 🥺 @andregasser

**Review by andregasser — APPROVED** [review]

**Pull request merged**